util/topsql/reporter/single_target.go

```diff
@@ -20,43 +20,97 @@ import (
 	"sync"
 	"time"
 
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/util"
 	"github.com/pingcap/tidb/util/logutil"
 	"github.com/pingcap/tipb/go-tipb"
 	"go.uber.org/zap"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/backoff"
 )
 
-// ReportClient send data to the target server.
-type ReportClient interface {
-	Send(ctx context.Context, addr string, data reportData) error
-	Close()
-}
-
-// GRPCReportClient reports data to grpc servers.
-type GRPCReportClient struct {
+// SingleTargetDataSink reports data to grpc servers.
+type SingleTargetDataSink struct {
 	curRPCAddr string
 	conn       *grpc.ClientConn
 	// calling decodePlan this can take a while, so should not block critical paths
 	decodePlan planBinaryDecodeFunc
+	sendTaskCh chan sendTask
 }
 
-// NewGRPCReportClient returns a new GRPCReportClient
-func NewGRPCReportClient(decodePlan planBinaryDecodeFunc) *GRPCReportClient {
-	return &GRPCReportClient{
+// NewSingleTargetDataSink returns a new SingleTargetDataSink
+func NewSingleTargetDataSink(decodePlan planBinaryDecodeFunc) *SingleTargetDataSink {
+	dataSink := &SingleTargetDataSink{
 		decodePlan: decodePlan,
+		sendTaskCh: make(chan sendTask, 1),
 	}
+	go util.WithRecovery(dataSink.run, nil)
```

Comment:
When panic happens, the data sink is no longer available. Is this behavior right?

Reply:
No, it should be available until `sendTaskCh` is closed. I am trying to figure out how to do it.

Reply:
PTAL @breeswish